apps/predictions/test/predictions_pub_sub_test.exs

```diff
@@ -0,0 +1,168 @@
+      predictions = [@prediction39]
+
+      :sys.replace_state(pid, fn state ->
+        Map.put(state, :predictions_by_route_id, %{@route_39 => predictions})
+      end)
+
+      assert PredictionsPubSub.subscribe(@route_39, pid) == predictions
+    end
+  end
+
+  describe "handle_info/2 - {:reset, predictions}" do
+    setup do
+      start_supervised({Registry, keys: :duplicate, name: :prediction_subscriptions_registry})
+
+      subscribe_fn = fn _, _ -> :ok end
+      {:ok, pid} = PredictionsPubSub.start_link(name: :subscribe, subscribe_fn: subscribe_fn)
+
+      :sys.replace_state(pid, fn state ->
+        Map.put(state, :predictions_by_route_id, %{@route_39 => [1, 2, 3]})
+      end)
+
+      {:ok, pid: pid}
+    end
+
+    test "resets the predictions", %{pid: pid} do
+      send(pid, {:reset, [@prediction39]})
+
+      assert pid |> :sys.get_state() |> Map.get(:predictions_by_route_id) |> Map.get(@route_39) ==
+               [
+                 @prediction39
+               ]
+    end
+
+    test "broadcasts new predictions lists to subscribers", %{pid: pid} do
+      PredictionsPubSub.subscribe(@route_39, pid)
+
+      send(pid, {:reset, [@prediction66]})
+      send(pid, {:reset, [@prediction39]})
+
+      assert_receive {:new_predictions, [@prediction39]}
+      refute_receive {:new_predictions, [@prediction66]}
+    end
+  end
+
+  describe "handle_info/2 - {:add, predictions}" do
+    setup do
+      start_supervised({Registry, keys: :duplicate, name: :prediction_subscriptions_registry})
+
+      subscribe_fn = fn _, _ -> :ok end
+      {:ok, pid} = PredictionsPubSub.start_link(name: :subscribe, subscribe_fn: subscribe_fn)
+
+      :sys.replace_state(pid, fn state ->
+        Map.put(state, :predictions_by_route_id, %{})
+      end)
+
+      {:ok, pid: pid}
+    end
+
+    test "adds the new predictions by route ID", %{pid: pid} do
+      send(pid, {:add, [@prediction39]})
+
+      assert pid |> :sys.get_state() |> Map.get(:predictions_by_route_id) ==
+               %{@route_39 => [@prediction39]}
+    end
+
+    test "broadcasts new predictions lists to subscribers", %{pid: pid} do
+      PredictionsPubSub.subscribe(@route_39, pid)
+
+      send(pid, {:add, [@prediction66]})
+      send(pid, {:add, [@prediction39]})
+
+      assert_receive {:new_predictions, [@prediction39]}
+      refute_receive {:new_predictions, [@prediction66]}
+    end
+  end
+
+  describe "handle_info/2 - {:update, predictions}" do
+    setup do
+      start_supervised({Registry, keys: :duplicate, name: :prediction_subscriptions_registry})
+
+      subscribe_fn = fn _, _ -> :ok end
+      {:ok, pid} = PredictionsPubSub.start_link(name: :subscribe, subscribe_fn: subscribe_fn)
+
+      :sys.replace_state(pid, fn state ->
+        Map.put(state, :predictions_by_route_id, %{@route_39 => [@prediction39]})
+      end)
+
+      {:ok, pid: pid}
+    end
+
+    test "updates the predictions", %{pid: pid} do
+      send(pid, {:update, [@prediction39]})
```

Comment:
**question:** am I reading this correctly that the initial state is `@prediction39` and it's being replaced with `@prediction39`? The test should probably take a modified prediction value to confirm that the update actually works.

Reply:
💯 hah, great catch! updated.